### PR TITLE
Kwabena/fix can

### PR DESCRIPTION
**Pull request by kwagyeman**

### CODE REVIEW

File: ports/stm32/fdcan.c [hunk 1]
```diff
@@ -199,27 +199,29 @@ void can_clearfilter(pyb_can_obj_t *self, uint32_t f, uint8_t bank) {
...
     } else {
         rxf = &can->Instance->RXF1S;
         rxa = &can->Instance->RXF1A;
+        sa = can->msgRam.RxFIFO1SA;
     }
 
     // Wait for a message to become available, with timeout
     uint32_t start = HAL_GetTick();
-    while ((*rxf & 7) == 0) {
+    while ((*rxf & FDCAN_RXF0S_F0FL) == 0) {
```
Comment (iabdalkader) [inline ports/stm32/fdcan.c]: Can you please use FDCAN_RXF1S_F1FL for FIFO1 ? Yes they're the same mask, but still. Could use a variable for the mask.

File: ports/stm32/fdcan.c [hunk 2]
```diff
@@ -199,27 +199,29 @@ void can_clearfilter(pyb_can_obj_t *self, uint32_t f, uint8_t bank) {
 }
 
 int can_receive(FDCAN_HandleTypeDef *can, int fifo, FDCAN_RxHeaderTypeDef *hdr, uint8_t *data, uint32_t timeout_ms) {
-    volatile uint32_t *rxf, *rxa;
+    volatile uint32_t *rxf, *rxa, sa;
```
Comment (iabdalkader) [inline ports/stm32/fdcan.c]: `sa` is not needed.. see the other comments.

File: ports/stm32/fdcan.c [hunk 3]
```diff
@@ -199,27 +199,29 @@ void can_clearfilter(pyb_can_obj_t *self, uint32_t f, uint8_t bank) {
...
         MICROPY_EVENT_POLL_HOOK
         if (HAL_GetTick() - start >= timeout_ms) {
             return -MP_ETIMEDOUT;
         }
     }
 
     // Get pointer to incoming message
-    uint32_t index = (can->Instance->RXF0S & FDCAN_RXF0S_F0GI) >> 8;
-    uint32_t *address = (uint32_t*)(can->msgRam.RxFIFO0SA + (index * can->Init.RxFifo0ElmtSize * 4));
+    uint32_t index = (*rxf & FDCAN_RXF0S_F0GI) >> FDCAN_RXF0S_F0GI_Pos;
```
Comment (iabdalkader) [inline ports/stm32/fdcan.c]: Please uses FIFO0 flags/mask for FIFO0 and FIFO1 flags/masks for FIFO1...
And use can->Init.RxFifoxElmtSize:
```
if (fifo == FDCAN_RX_FIFO0) {
index = (*rxf & FDCAN_RXF0S_F0GI) >> FDCAN_RXF0S_F0GI_Pos;
address = (uint32_t*)(can->msgRam.RxFIFO0SA + (index * can->Init.RxFifo0ElmtSize * 4));
} else {
index = (*rxf & FDCAN_RXF1S_F1GI) >> FDCAN_RXF1S_F1GI_Pos;
address = (uint32_t*)(can->msgRam.RxFIFO1SA + (index * can->Init.RxFifo1ElmtSize * 4));
}
```

**Review by iabdalkader — CHANGES_REQUESTED** [review]

File: ports/stm32/pyb_can.c [hunk 4]
```diff
@@ -423,6 +423,18 @@ STATIC mp_obj_t pyb_can_send(size_t n_args, const mp_obj_t *pos_args, mp_map_t *
 
     HAL_StatusTypeDef status;
     #if MICROPY_HW_ENABLE_FDCAN
+    uint32_t Timeout = args[ARG_timeout].u_int;
```
Comment (iabdalkader) [inline ports/stm32/pyb_can.c]: please rename Timeout to timeout_ms.

Comment (kwagyeman) [reply ports/stm32/pyb_can.c]: Fixed.

File: ports/stm32/fdcan.c (hunk 1 above)
Comment (kwagyeman) [reply ports/stm32/fdcan.c]: Done

File: ports/stm32/fdcan.c (hunk 3 above)
Comment (kwagyeman) [reply ports/stm32/fdcan.c]: Done

File: ports/stm32/fdcan.c [hunk 5]
```diff
@@ -200,26 +200,40 @@ void can_clearfilter(pyb_can_obj_t *self, uint32_t f, uint8_t bank) {
...
+        fl = FDCAN_RXF1S_F1FL;
     }
 
     // Wait for a message to become available, with timeout
     uint32_t start = HAL_GetTick();
-    while ((*rxf & 7) == 0) {
-        MICROPY_EVENT_POLL_HOOK
-        if (HAL_GetTick() - start >= timeout_ms) {
+    while ((*rxf & fl) == 0) {
+        if (timeout_ms == 0) {
```
Comment (iabdalkader) [inline ports/stm32/fdcan.c]: I think timeout_ms == 0 means No timeout.. This loop should stay the same except for fixing the flag.

File: ports/stm32/pyb_can.c [hunk 6]
```diff
@@ -423,6 +423,20 @@ STATIC mp_obj_t pyb_can_send(size_t n_args, const mp_obj_t *pos_args, mp_map_t *
 
     HAL_StatusTypeDef status;
     #if MICROPY_HW_ENABLE_FDCAN
+    uint32_t timeout_ms = args[ARG_timeout].u_int;
```
Comment (iabdalkader) [inline ports/stm32/pyb_can.c:426]: The behavior of timeout is actually documented [here](http://docs.micropython.org/en/latest/library/pyb.CAN.html?highlight=can#pyb.CAN.send):
`If timeout is 0 the message is placed in a buffer in one of three hardware buffers and the method returns immediately. If all three buffers are in use an exception is thrown. If timeout is not 0, the method waits until the message is transmitted. If the message can’t be transmitted within the specified time an exception is thrown.`

**Review by iabdalkader — CHANGES_REQUESTED** [review]

**Comment by kwagyeman** [comment]
It matches now.

**Pull request merged**